adapting the script movielens_recommendations_transformers.py to be Backend-Agnostic

Humbulani1234 opened the pull request:
This PR adapts the script ` movielens_recommendations_transformers.py ` to be ***Backend-Agnostic***

fchollet commented on examples/structured_data/movielens_recommendations_transformers.py:
```diff
@@ -292,198 +353,206 @@ def process(features):
+        encoded_other_features = []
+
+        # Helper function to create embeddings
+        def embedding_helper(input):
```
Please move this to be a method on the class

Humbulani1234 replied:
Moved.

fchollet commented on examples/structured_data/movielens_recommendations_transformers.py:
```diff
@@ -292,198 +353,206 @@ def process(features):
+
+            # This function merely groups similar logic for include_movie_features=True,
+            # or given as include_movie_features=False
+            def movie_sequence_helper(encoded_sequence_movies):
```
Same here

Humbulani1234 replied:
Moved

fchollet commented on examples/structured_data/movielens_recommendations_transformers.py:
```diff
@@ -292,198 +353,206 @@ def process(features):
+            # or given as include_movie_features=False
+            def movie_sequence_helper(encoded_sequence_movies):
+                # Create positional embedding.
+                positions = keras.ops.arange(start=0, stop=sequence_length - 1, step=1)
```
Instead of using `keras.ops.*` everywhere, just do `from keras import ops` at the start -- it will shorten many lines

fchollet commented on examples/structured_data/movielens_recommendations_transformers.py:
```diff
@@ -292,198 +353,206 @@ def process(features):
-        output_dim=movie_embedding_dims,
-        name="position_embedding",
-    )
-    positions = tf.range(start=0, limit=sequence_length - 1, delta=1)
```
Did the old really need extensive refactoring? Best I can tell only this line needed to change (to become `ops.arange()`

Humbulani1234 replied:
Changing only the `tf.range` to `ops.arange` does not work and I believe it is beacuse `StringLookup` is part of the model, not of `tf.data.Dataset`. 

So, my approach was to split the function `encode_input_features` into two parts `StringLookups` and `Embeddings`. I placed the `StringLookups` functionality into `tf.data.Dataset` processing step, and created a class to handle each input `embeddings`. And also the issue that this script must run whether one chooses to include `user_features` or not also adds some refactoring.

However, if there is a more less-refactoring approach/method, I'm wiling to learn and implement it. I must confess also that the refactoring felt extensive.

fchollet reviewed (commented):
Thanks for the PR!

Humbulani1234 commented:
PR addressing comments.

Humbulani1234 commented:
Indeed, there was extensive refactoring. I've implemented the following new approach:

- Kept the old function `encode_input_features`; and
- Removed only the `StringLookups` functionality from the above function and deploy it into `tf.data.Dataset` processing.

This PR implements the changes

divyashreepathihalli commented:
Thanks @Humbulani1234! LGTM,  can you please generate the .ipynb and .md files?

Humbulani1234 commented:
Generated `.md` and `.ipynb` files.
